### ISSUE TITLE
Support ES6 constructor.name 

### ISSUE THREAD

**Issue by mikemaccana** [microsoft/TypeScript#37340]
**TypeScript Version:**  Nightly

**Search Terms:** 'constuctor name'

**Expected behavior:**
No errors.

**Actual behavior:**
See the playground and code links below. Uncomment the notes line and see that typescript produces z Property 'join' does not exist on type 'string'.` for code that only runs on Arrays.

**Related Issues:**
Nope

**Code**
```ts
function runMe(options: any) {
  const params = {};
  Object.keys(options).forEach(function (option: string | string[]) {
    // Using the ES6 constructor.name per below will trigger incorrect warnings
    // if (option.constructor.name === 'Array') {
    if (Array.isArray(option)) {
      option = option.join(",")
    }
  });
  return params;
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function runMe(options) {
    const params = {};
    Object.keys(options).forEach(function (option) {
        // Using the ES6 cnstructor.name per below will trigger 
        // if (option.constructor.name === 'Array') {
        if (Array.isArray(option)) {
            option = option.join(",");
        }
    });
    return params;
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](http://www.typescriptlang.org/play/?ts=3.9.0-dev.20200310&ssl=1&ssc=10&pln=1&pc=15#)

**Comment by RyanCavanaugh** [comment]
?
![image](https://user-images.githubusercontent.com/6685088/76470397-436e1000-63ad-11ea-8a3f-8232b792b2c1.png)

**Comment by mikemaccana** [comment]
![image](https://user-images.githubusercontent.com/172594/76542315-b5615a80-647c-11ea-80a7-e377b5f6be7e.png)

**Comment by RyanCavanaugh** [comment]
This isn't a type guard pattern we support, nor do we intend to.

**Comment by mikemaccana** [comment]
Why not? `constructor.name` is [part of ES6](http://www.ecma-international.org/ecma-262/6.0/#sec-runtime-semantics-bindingclassdeclarationevaluation), and allows users to determine type more easily than knowing each type-specific check. Here's a quote from [Domenic Denicola](https://stackoverflow.com/questions/29310530/get-the-class-name-of-es6-class-instance):

> The name property of functions declared via ClassDeclaration productions is set in 14.5.15, step 6.

**Comment by RyanCavanaugh** [comment]
This kind of check gets it wrong in both directions. You could have a value whose `constructor.name` property was `"Array"` that *wasn't* an array (because there was a `class Array` declared somewhere else), and you could have a value whose `constructor.name` *wasn't* `"Array"` but *was* an array (because it's e.g. `class MyArray extends Array`).

**Comment by mikemaccana** [comment]
Hi @RyanCavanaugh! Understood and thanks for the reply. 👍🏽
